tests/unit/ZeroBounceValidatorTest.php

```diff
@@ -0,0 +1,172 @@
+use alexeevdv\yii\zerobounce\TransportException;
+use alexeevdv\yii\zerobounce\ValidateResponse;
+use alexeevdv\yii\zerobounce\ZeroBounceValidator;
+use Codeception\Stub\Expected;
+use Codeception\Test\Unit;
+use Yii;
+use yii\console\Request as ConsoleRequest;
+use yii\httpclient\Response;
+use yii\web\Request;
+
+class ZeroBounceValidatorTest extends Unit
+{
+
+    /**
+     * @dataProvider validResponses
+     */
+    public function testValidEmail($response)
+    {
+        $validator = new ZeroBounceValidator([
+            'client' => $this->make(Client::class, [
+                'validate' => Expected::once(function ($value, $ip) use ($response) {
+                    $this->assertEquals('test@test.com', $value);
+
+                    return new ValidateResponse([
+                        'status' => $response
+                    ]);
+                })
+            ])
+        ]);
+
+        $this->assertTrue($validator->validate('test@test.com'));
+    }
+
+    public function validResponses()
+    {
+        return [
+            'valid' => ['valid'],
+            'unknown' => ['unknown'],
+        ];
+    }
+
+    /**
+     * @dataProvider invalidResponses
+     */
+    public function testInalidEmail($response)
+    {
+        $validator = new ZeroBounceValidator([
+            'client' => $this->make(Client::class, [
+                'validate' => Expected::once(function ($value, $ip) use ($response) {
+                    $this->assertEquals('test@test.com', $value);
+                    $this->assertEmpty($ip);
+
+                    return new ValidateResponse([
+                        'status' => $response
+                    ]);
+                })
+            ])
+        ]);
+
+        $this->assertFalse($validator->validate('test@test.com'));
+    }
+
+    public function invalidResponses()
+    {
+        return [
+            'invalid' => ['invalid'],
+            'catch-all' => ['catch-all'],
+            'spamtrap' => ['spamtrap'],
+            'abuse' => ['abuse'],
+            'do_not_mail' => ['do_not_mail'],
+        ];
+    }
+
+    public function testExtractedIpAddress()
+    {
+        Yii::$app->setComponents([
+            'request' => $this->make(Request::class, [
+                'getUserIP' => Expected::once('127.0.0.1')
+            ])
+        ]);
+
+        $validator = new ZeroBounceValidator([
+            'client' => $this->make(Client::class, [
+                'validate' => Expected::once(function ($value, $ip) {
+                    $this->assertEquals('test@test.com', $value);
+                    $this->assertEquals('127.0.0.1', $ip);
+
+                    return new ValidateResponse([
+                        'status' => 'valid'
+                    ]);
+                })
+            ])
+        ]);
+
+        $this->assertTrue($validator->validate('test@test.com'));
+    }
+
+    public function testNotWebRequest()
+    {
+        Yii::$app->setComponents([
```

Comment:
This is not a proper unit test. You should not configure global state. You need to configure module you are testing. Give it what it needs as a dependency

Reply:
there is no way to do it in yii2, you need to cofigure framework itself.. But in unit tests it is not global(that is why Yii2 module is needed, as it initializes new framework for each test)

Reply:
It has nothing to do with yii2. You can make it work by proper dependency injection. Just dont hardcode stuff like `Yii::$app->request->whatever()`

Reply:
no global state usage => no need to configure global state